promql/engine.go

```diff
@@ -576,45 +579,137 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
 	return mat, warnings, nil
 }
 
-// subqueryOffsetRange returns the sum of offsets and ranges of all subqueries in the path.
-func (ng *Engine) subqueryOffsetRange(path []parser.Node) (time.Duration, time.Duration) {
+// subqueryTimes returns the sum of offsets and ranges of all subqueries in the path.
+// It also returns the min and max timestamp associated with the subquery.
+// If the @ modifier is used, then the offset is w.r.t. the startTime passed.
+func subqueryTimes(path []parser.Node) (time.Duration, time.Duration, int64, int64) {
 	var (
-		subqOffset time.Duration
-		subqRange  time.Duration
+		subqOffset, subqRange      time.Duration
+		minTimestamp, maxTimestamp int64 = math.MaxInt64, math.MinInt64
 	)
 	for _, node := range path {
 		switch n := node.(type) {
 		case *parser.SubqueryExpr:
-			subqOffset += n.Offset
+			subqOffset += n.OriginalOffset // TODO(codesome): fix this calculation
 			subqRange += n.Range
+			if n.Timestamp != nil {
+				if *n.Timestamp < minTimestamp {
+					minTimestamp = *n.Timestamp
+				}
+				if *n.Timestamp > maxTimestamp {
+					maxTimestamp = *n.Timestamp
+				}
+			}
 		}
 	}
-	return subqOffset, subqRange
+	return subqOffset, subqRange, minTimestamp, maxTimestamp
 }
 
-func (ng *Engine) findMinTime(s *parser.EvalStmt) time.Time {
+func (ng *Engine) findMinMaxTime(s *parser.EvalStmt) (int64, int64) {
```

Comment:
This doesn't feel right with subqueries, for example subqMint is not the closest timestamp subquery but the minimum.

Reply:
Right